src/server/game/Movement/PathGenerator.cpp

```diff
@@ -622,8 +622,34 @@ void PathGenerator::BuildPointPath(const float *startPoint, const float *endPoin
 
 void PathGenerator::NormalizePath()
 {
-    for (uint32 i = 0; i < _pathPoints.size(); ++i)
-        _source->UpdateAllowedPositionZ(_pathPoints[i].x, _pathPoints[i].y, _pathPoints[i].z);
+    constexpr float MAX_Z_DELTA_THRESHOLD = 4.0f;
+
+    bool shouldRespectThreshold = [this]
+    {
+        Unit const* sourceUnit = _source->ToUnit();
+        if (!sourceUnit)
+            return false;
+
+        if (sourceUnit->IsFlying() || sourceUnit->IsInWater())
+            return false;
+
+        if (!(_type & PATHFIND_NORMAL))
+            return false;
+
+        return true;
+    }();
+
+    for (G3D::Vector3& pathPoint : _pathPoints)
+    {
+        float previousZ = pathPoint.z;
+
+        _source->UpdateAllowedPositionZ(pathPoint.x, pathPoint.y, pathPoint.z);
```

Comment:
another problem are the transitions (if a creature is not in water now, but it will be when path node is reached, creature is not flying now, but will be when node is reaching, creature is not hovering now...)

It'd be far more secure if this is only applied on creatures that will not have these transitions and strictly can only walk on ground (canswim canfly canhover...), and still you are "screwed up" if a script changes anything...

Reply:
but isn't that how the current behavior is and any change to that should be out of scope from this PR ?